ls-ecaps.c

```diff
@@ -690,30 +690,206 @@ cap_rcec(struct device *d, int where)
 }
 
 static void
-cap_dvsec_cxl(struct device *d, int where)
+cxl_range(u64 base, u64 size, int n)
 {
-  u16 l;
+  u32 interleave[] = { 0, 256, 4096, 512, 1024, 2048, 8192, 16384 };
+  const char *type[] = { "Volatile", "Non-volatile", "CDAT" };
+  const char *class[] = { "DRAM", "Storage", "CDAT" };
+  u16 w;
 
-  printf(": CXL\n");
-  if (verbose < 2)
-    return;
+  w = (u16) size;
+
+  size &= ~0x0fffffffULL;
+
+  printf("\t\tRange%d: %016"PCI_U64_FMT_X"-%016"PCI_U64_FMT_X"\n", n, base, base + size - 1);
+  printf("\t\t\tValid%c Active%c Type=%s Class=%s interleave=%d timeout=%ds\n",
+    FLAG(w, PCI_CXL_RANGE_VALID), FLAG(w, PCI_CXL_RANGE_ACTIVE),
+    type[PCI_CXL_RANGE_TYPE(w)], class[PCI_CXL_RANGE_CLASS(w)],
+    interleave[PCI_CXL_RANGE_INTERLEAVE(w)],
+    1 << (PCI_CXL_RANGE_TIMEOUT(w) * 2));
+}
 
-  if (!config_fetch(d, where + PCI_CXL_CAP, 12))
+static void
+dvsec_cxl_device(struct device *d, int where, int rev)
+{
+  u32 cache_size, cache_unit_size, l;
+  u64 range_base, range_size;
+  u16 w;
+
+  /* Legacy 1.1 revs aren't handled */
+  if (rev < 1)
     return;
```

Comment:
How is mapped `int rev` to the revision tuple major.minor? Because this check reject rev==0 but comment say that 1.1. So seems like there is non-trivial mapping between integer and tuple major.minor (if it is really major.minor format).

Reply:
From 8.1.3 of the CXL 3.0 specification, the DVSEC Revision ID is mapped as follows: 
`
DVSEC Revision ID 0h represents the structure as defined in the CXL 1.1 specification.
`
`
DVSEC Revision ID 1h represents the structure as defined in the CXL 2.0 specification.
`
`
DVSEC Revision ID 2h represents the structure as defined in the CXL 3.0 specification.
`

Reply:
Thanks! Interesting mapping... and code with comment looks correct.